### PR TITLE
Adds read locks to LuceneSearchProvider.

### PR DESCRIPTION
**Pull request by archolewa**
The `LuceneSearchProvider` is missing read locks before reading the lucene database. As a result, Bad Things can happen if you attempt to read from Lucene (i.e. resolving dimension filters) while loading dimension indices.

### CODE REVIEW

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java [hunk 1]
```diff
@@ -547,17 +554,23 @@ private Query getFilterQuery(Set<ApiFilter> filters, int perPage) {
                 .map(document -> document.get(idKey))
                 .map(dimension::findDimensionRowByKeyValue)
                 .collect(Collectors.toCollection(TreeSet::new));
+
+        int documentCount;
+        lock.readLock().lock();
```
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java]: Don't we need to get this lock much earlier in this method? We're getting `indexSearcher` up near the top of the method...

**Comment by michael-mclawhorn** [comment]
What would be great would be if we had some way of testing access to the indexer low levels so we could validate that all calls to them were only done in a readLocked context.

It looks pretty good.  Needs a CHANGELOG.

**Comment by cdeszaq** [comment]
A few more thoughts:
-Probably also need a write-lock in the `reopenIndexSearcher` method when we close/re-open it
- `clearDimension` needs some help:
  -It currently will deadlock, since it holds a write-lock, and calls `refreshCardinality` from within it, which tries to grab a read-lock. I think moving the method calls to`refreshCardinality` and `reopenIndexSearcher` outside the `try/catch/finally` is all that's needed, but I didn't trace it too carefully

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java (hunk 1 above)
Comment (archolewa) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java]: Yes.

**Comment by archolewa** [comment]
@michael-mclawhorn You _might_ be able to do something by taking a write lock in a separate thread and then inspecting the guts of the read-write lock. I did a little bit of digging, but it's not immediately clear to me how to do that.

Alternatively, you could take a write lock and then verify that the method doesn't finish, but that would significantly increase our testing time.

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java [hunk 2]
```diff
@@ -513,51 +542,57 @@ private Query getFilterQuery(Set<ApiFilter> filters, int perPage) {
...
-            }
-        }
-        for (int currentPage = 1; currentPage < requestedPageNumber; currentPage++) {
-            ScoreDoc lastEntry = hits[hits.length - 1];
-            hits = getPageOfData(indexSearcher, lastEntry, query, perPage, requestedPageNumber).scoreDocs;
+        TreeSet<DimensionRow> filteredDimRows;
+        int documentCount;
+        initializeIndexSearcher();
+        try {
+            lock.readLock().lock();
```
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java]: Move outside the `try` to match the other calls in this class?

**Comment by cdeszaq** [comment]
👍

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java (hunk 2 above)
Comment (wcekan) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java]: Correct.  any lock must complete before the `try`

**Comment by wcekan** [comment]
👍

**Pull request merged**